src/ReservoirDeployer.sol

```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.0;
+
+import { Address } from "@openzeppelin/utils/Address.sol";
+
+import { FactoryStoreLib } from "src/libraries/FactoryStore.sol";
+
+import { OracleCaller } from "src/oracle/OracleCaller.sol";
+import { GenericFactory } from "src/GenericFactory.sol";
+
+// TODO:
+// - Enable factory ownership transfer.
+// - Enable oracle caller ownership transfer.
+
+contract ReservoirDeployer {
+    using FactoryStoreLib for GenericFactory;
+
+    // Default configuration.
+    uint256 public constant INITIAL_MINT_AMOUNT = 100e18;
+    uint256 public constant DEFAULT_SWAP_FEE_CP = 3000; // 0.3%
+    uint256 public constant DEFAULT_SWAP_FEE_SP = 100; // 0.01%
+    uint256 public constant DEFAULT_PLATFORM_FEE = 250_000; // 25%
+    uint256 public constant DEFAULT_AMP_COEFF = 1000;
+    uint256 public constant DEFAULT_MAX_CHANGE_RATE = 0.0005e18;
+
+    // Steps.
+    uint256 public constant TERMINAL_STEP = 4;
+    uint256 public step = 0;
+
+    // Bytecode hashes.
+    bytes32 public constant factory_hash = bytes32(0x9f1cc554aa194d48ae1fe96a6f5ea41fbabfaa04128fd91f6ec1f290f8756324);
+    bytes32 public constant constant_product_hash =
+        bytes32(0xc3c98a193686624a579c5c3324bf9aff1fc419aaa1c47076b76fd381b90d27cb);
+    bytes32 public constant stable_hash = bytes32(0x133955ed8f42488ffb022fe8aa4882b1a2e05daca08c7affb2be519551dc07ef);
+    bytes32 public constant oracle_caller_hash =
+        bytes32(0x262458524d9c8928fe7fd7661236b93f6d6a9535182f48fd582a75f18bfbf85f);
+
+    // Deployment addresses.
+    GenericFactory public factory;
+    OracleCaller public oracleCaller;
+
+    function isDone() external view returns (bool) {
+        return step == TERMINAL_STEP;
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                            DEPLOYMENT STEPS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    function deployFactory(bytes memory aFactoryBytecode) external returns (GenericFactory) {
+        require(step == 0, "FAC_STEP: OUT_OF_ORDER");
+        require(keccak256(aFactoryBytecode) == factory_hash);
+
+        // Manual deployment from validated bytecode.
+        address lFactoryAddress;
+        assembly ("memory-safe") {
+            lFactoryAddress :=
+                create(
+                    0, // value
+                    add(aFactoryBytecode, 0x20), // offset
+                    mload(aFactoryBytecode) // size
+                )
+        }
+        require(lFactoryAddress != address(0), "FAC_STEP: DEPLOYMENT_FAILED");
+
+        // Write the factory address so we can start configuring it.
+        factory = GenericFactory(lFactoryAddress);
+
+        // Set global parameters.
+        factory.write("Shared::platformFee", DEFAULT_PLATFORM_FEE);
+        factory.write("Shared::platformFeeTo", address(this));
+        factory.write("Shared::recoverer", msg.sender);
```

Comment:
do we want to set these two OUTSIDE of the deployer? Cuz we may not know which addresses these are at the time of factory deployment. We certainly don't want the `platformFeeTo` to be the deployer do we?

Reply:
If you mentally rename "deployer" to "crosschain multisig", does it then seem fine to set these?

Reply:
yes if it's a cross-chain multi-sig then it's fine

Reply:
Okay, that's essentially what the guardians + rawCall enable